src/utils/api-request.ts

```diff
@@ -17,15 +17,293 @@
+      } else if (validator.isBuffer(config.data)) {
+        data = config.data as Buffer;
+      } else {
+        return reject(createError(
+          'Request data must be a string, a Buffer or a json serializable object',
+          config,
+        ));
+      }
+      // Add Content-Length header if data exists
+      headers['Content-Length'] = data.length.toString();
+    }
+    const parsed = url.parse(config.url);
+    const protocol = parsed.protocol || 'https:';
+    const isHttps = protocol === 'https:';
+    const options = {
+      hostname: parsed.hostname,
+      port: parsed.port,
+      path: parsed.path,
+      method: config.method,
+      headers,
+    };
+    const transport: any = isHttps ? https : http;
+    const req: http.ClientRequest = transport.request(options, (res: http.IncomingMessage) => {
+      if (req.aborted) {
+        return;
+      }
+      // uncompress the response body transparently if required
+      let respStream: stream.Readable = res;
+      const encodings = ['gzip', 'compress', 'deflate'];
+      if (encodings.indexOf(res.headers['content-encoding']) !== -1) {
+        // add the unzipper to the body stream processing pipeline
+        const zlib: typeof zlibmod = require('zlib');
+        respStream = respStream.pipe(zlib.createUnzip());
+        // remove the content-encoding in order to not confuse downstream operations
+        delete res.headers['content-encoding'];
+      }
+
+      const response: LowLevelResponse = {
+        status: res.statusCode,
+        headers: res.headers,
+        request: req,
+        data: undefined,
+        config,
+      };
+
+      const responseBuffer = [];
+      respStream.on('data', (chunk) => {
+        responseBuffer.push(chunk);
+      });
+
+      respStream.on('error', (err) => {
+        if (req.aborted) {
+          return;
+        }
+        reject(enhanceError(err, config, null, req));
+      });
+
+      respStream.on('end', () => {
+        const responseData = Buffer.concat(responseBuffer).toString();
+        response.data = responseData;
+        settle(resolve, reject, response);
+      });
+    });
+
+    // Handle errors
+    req.on('error', (err) => {
+      if (req.aborted) {
+        return;
+      }
+      reject(enhanceError(err, config, null, req));
+    });
+    if (config.timeout) {
+      // Listen to timeouts and throw an error.
+      req.setTimeout(config.timeout, () => {
+        req.abort();
+        reject(createError(`timeout of ${config.timeout}ms exceeded`, config, 'ETIMEDOUT', req));
+      });
+    }
+    // Send the request
+    req.end(data);
+  });
+}
+
+function createError(
+  message: string,
+  config: HttpRequestConfig,
+  code?: string,
+  request?: http.ClientRequest,
+  response?: LowLevelResponse): LowLevelError {
+
+  const error = new Error(message);
+  return enhanceError(error, config, code, request, response);
+}
+
+function enhanceError(
+  error,
```

Comment:
error: Error,

Reply:
That's not possible since we add new fields to the error object in the method body (triggers TS compilation error).